### PR TITLE
💻 : arcee cli

### PR DESCRIPTION
**Pull request by EricLiclair**
I'm considering the implementation of a CLI (Command-Line Interface) for quick validation purposes. The envisioned CLI would be utilized as follows:

```shell
pip install arcee-py

# To generate content
arcee generate

# To retrieve content
arcee retrieve
```

The primary use case for this CLI is to interact with a generative language model. While I initially had some uncertainty regarding the input format, it appears that using a string as input should suffice.

I would greatly appreciate your thoughts and feedback on this proposal, @Jacobsolawetz. Please let me know if you have any suggestions or concerns.

image

### CODE REVIEW

**Comment by Jacobsolawetz** [comment]
@EricLiclair That sounds like a great idea - would you mind if we hold off until the API is a bit more solidified and then I will reach out to you here on this thread?

**Comment by Ben-Epstein** [comment]
@EricLiclair this is great, thank you! 

Would you mind building this with typer instead of click? It should be (almost) as simple as `:s/click/typer/g`

Typer gives us a few extras without being much heavier that, if we're starting from scratch, I'd love to get out of the box (type hints for validation, auto cli docs, cli autocompletion).

**Comment by EricLiclair** [comment]
Hey, @Jacobsolawetz I don't mind at all and sure. If I'm unresponsive you can reach out to me via linkedin or through Mark. 😊

Hi, @Ben-Epstein, Insightful inputs. I've added an exact replica of the previous implementation using typer. 
With a little better understanding of the api structure there's always room for a better implementation.

_An example impl. for variables as command arguments, etc. Nonetheless, let me know if you'd need anything else_
```python
@cli.command()
def retrieve(model_name: str = typer.Argument(..., help="Model name")):
    """Retrieve from API"""

    dalm = DALM(name=model_name)
    typer.echo(
        typer.style(
            f"You are now configured with Dalm - model {model_name}",
            fg=typer.colors.GREEN,
        )
    )
...
```

_which would be used like:_
image

**Comment by Ben-Epstein** [comment]
@EricLiclair fantastic! That's exactly what I was planning on doing next week :)

I think that we'll likely want to support passing a text file or jsonl, for commands like `arcee upload context new_context -f my data.jsonl` 

But that doesn't need to be now, getting a solid design in place is great for now. 

This is perfect.

File: arcee/cli.py
Comment (Ben-Epstein) [inline arcee/cli.py:71]:
```suggestion
    if file.exists():
```

**Comment by Ben-Epstein** [comment]
Hey @EricLiclair Thanks for updating after our merge yesterday. I think to get this CLI to a really useful spot, we need 3 more things.

1. We need an `upload contexts` so that a user can upload many files at once (it can call `[upload_docs](https://github.com/arcee-ai/arcee-python/blob/main/arcee/api.py#L35)`). It can take a directory to files, and load them all and upload
2. A `arcee train` command, so we can close the loop
3. a simple section in the readme showcasing an easy way to achieve an e2e train

for (3), something like 
```markdown
## Using the CLI

```shell
# upload context data to arcee
arcee upload contexts context1 --file  /path/to/documents
arcee train dalm1 --context context1
# ... wait for training to complete
arcee retrieve --name dalm1 --query "what is the capital of Washington State?"
arcee generate --name dalm1 --query "what is the capital of Washington State?"
\```
```

File: arcee/cli.py
Comment (EricLiclair) [reply arcee/cli.py]: Thanks for the comment. I've replaced `os` module with `pathlib` for all validation and read operations.

**Comment by Ben-Epstein** [comment]
@EricLiclair 

> iteratively hit the api with data from k files at a time, k < n but does the service allow us to upload multiple docs to the same context and merge them?

Yes this is perfect. And yes we support uploading multiple docs to the same context! 

> File validations. What files are valid? currently validating if extension is one of .txt or .jsonl

Thats perfect!

**Comment by EricLiclair** [comment]
Hey, @Ben-Epstein 
Apologies for the delayed PR. 
Have updated the upload logic to check file sizes and upload in chunks/buffers; defaults to 512 MB.

**Review by Ben-Epstein — APPROVED** [review]
great! thanks for this awesome contribution @EricLiclair

**Pull request merged**
